Rate limit GC

Andrew-Fryer opened the pull request:
#### 7417bace367861c80a07197c1a7535c1620e3b51
<pre>
Rate limit GC
<a href="https://bugs.webkit.org/show_bug.cgi?id=295839">https://bugs.webkit.org/show_bug.cgi?id=295839</a>
<a href="https://rdar.apple.com/155699898">rdar://155699898</a>

Reviewed by Yusuke Suzuki.

For machines with &gt;= 16 GB RAM, accumulate the number of GCs we&apos;ve done
with a given half-life to determine if the heap has been GCing a lot.
If so, don&apos;t trigger GC in slow allocation until we&apos;ve reached another
threshold after m_maxHeapSize, allowing eden to grow to a constant
multiple of the minimum heap size.

This won&apos;t increase memory usage too much or unecessarily because:
The threshold prevents eden from increasing memory usage too much.
Exponential decay means we only do this if we&apos;ve been GCing more than
once a half-life recently.

Also adjust js stress tests so that this doesn&apos;t break them.
(Increase js stress test memory limit and add explicit gc calls to a
test that ensures substrings of rope don&apos;t use too much memory.)

* JSTests/stress/many-substrings-of-rope-shouldnt-use-excessive-memory-2.js:
(test):
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::Heap):
(JSC::Heap::runEndPhase):
(JSC::Heap::projectedGCRateLimitingValue):
(JSC::Heap::updateAllocationLimits):
(JSC::Heap::collectIfNecessaryOrDefer):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Tools/Scripts/run-jsc-stress-tests:

Canonical link: <a href="https://commits.webkit.org/297631@main">https://commits.webkit.org/297631@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/51bd948d58ec7dc4876019ef8a35ecbb10bf6ece

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112362 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32094 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22571 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118441 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62734 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114324 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32746 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40657 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85357 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/36063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115309 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26135 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101092 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65788 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25433 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19228 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62286 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/104876 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95519 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19306 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121767 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/110976 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39436 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29359 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94165 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39817 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97333 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93991 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24032 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39237 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17031 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/35486 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39324 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44812 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/135207 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38959 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36332 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42296 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40702 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Andrew-Fryer commented on Source/JavaScriptCore/heap/Heap.cpp:
```diff
@@ -2854,6 +2880,12 @@ void Heap::collectIfNecessaryOrDefer(GCDeferralContext* deferralContext)
             return false;
+        if (m_maxEdenSizeForRateLimiting) {
+            if (m_sizeAfterLastCollect + bytesAllocatedThisCycle <= m_sizeBeforeLastFullCollect || bytesAllocatedThisCycle < m_maxEdenSizeForRateLimiting) {
+                if (projectedGCRateLimitingVal(MonotonicTime::now()) > 1.0)
```
If the value is > 1.0, then we've been GCing more frequently than once every half-life.

Constellation commented on Source/JavaScriptCore/heap/Heap.h:
```diff
@@ -815,6 +816,8 @@ class Heap {
     const size_t m_ramSize;
     const size_t m_minBytesPerCycle;
+    const size_t m_maxEdenSizeForRateLimiting;
+    double m_gcRateLimitingVal { 0.0 };
```
I would like to say `Value` instead of `Val`.

Constellation commented on Source/JavaScriptCore/heap/Heap.cpp:
```diff
@@ -138,6 +138,16 @@ size_t minHeapSize(HeapType heapType, size_t ramSize)
+    // Only do rate limiting for machines >= 16GB RAM.
+    size_t aggressiveHeapThresholdInBytes = static_cast<size_t>(Options::aggressiveHeapThresholdInMB()) * MB;
+    if (ramSize < aggressiveHeapThresholdInBytes)
+        return 0;
```
Maybe we should factor this condition check out, and place it into a member of `Heap` as enum, like,
`enum class GrowthMode : uint8_t { Default, Aggressive };`
and
`GrowthMode m_growthMode { ... }`
And passing this enum to this function instead of computing this in different places multiple times.

Constellation commented on Source/JavaScriptCore/heap/Heap.cpp:
```diff
@@ -147,8 +157,8 @@ size_t proportionalHeapSize(size_t heapSize, size_t ramSize)
-    size_t heapGrowthFunctionThresholdInBytes = static_cast<size_t>(Options::heapGrowthFunctionThresholdInMB()) * MB;
-    if (ramSize < heapGrowthFunctionThresholdInBytes)
+    size_t aggressiveHeapThresholdInBytes = static_cast<size_t>(Options::aggressiveHeapThresholdInMB()) * MB;
+    if (ramSize < aggressiveHeapThresholdInBytes)
```
Ditto for that refactoring.

Constellation commented on Source/JavaScriptCore/heap/Heap.cpp:
```diff
@@ -2437,6 +2453,16 @@ void Heap::notifyIncrementalSweeper()
     m_sweeper->startSweeping(*this);
 }
 
+double Heap::projectedGCRateLimitingVal(MonotonicTime now)
```
Let's say `Value` instead of `Val`.

Constellation commented on Source/JavaScriptCore/heap/Heap.cpp:
```diff
@@ -138,19 +148,23 @@ size_t minHeapSize(HeapType heapType, size_t ramSize)
+    return 0.0;
+}
+
+size_t proportionalHeapSize(size_t heapSize, GrowthMode growthMode, size_t ramSize)
```
Let's put `static`

Constellation commented on Source/JavaScriptCore/heap/Heap.cpp:
```diff
@@ -138,19 +148,23 @@ size_t minHeapSize(HeapType heapType, size_t ramSize)
 }
 
-size_t proportionalHeapSize(size_t heapSize, size_t ramSize)
+size_t maxEdenSizeForRateLimiting(GrowthMode growthMode, size_t minBytesPerCycle)
```
Let's put `static`

Constellation commented on Source/JavaScriptCore/heap/Heap.cpp:
```diff
@@ -121,6 +121,16 @@ double maxPauseMS(double thisPauseMS)
     return maxPauseMS;
 }
 
+GrowthMode growthMode(size_t ramSize)
```
Let's put `static`

Constellation reviewed (approved):
r=me with a bit comments.

Andrew-Fryer committed:
Rate limit GC
https://bugs.webkit.org/show_bug.cgi?id=295839
rdar://155699898

Reviewed by Yusuke Suzuki.

For machines with >= 16 GB RAM, accumulate the number of GCs we've done
with a given half-life to determine if the heap has been GCing a lot.
If so, don't trigger GC in slow allocation until we've reached another
threshold after m_maxHeapSize, allowing eden to grow to a constant
multiple of the minimum heap size.

This won't increase memory usage too much or unecessarily because:
The threshold prevents eden from increasing memory usage too much.
Exponential decay means we only do this if we've been GCing more than
once a half-life recently.

Also adjust js stress tests so that this doesn't break them.
(Increase js stress test memory limit and add explicit gc calls to a
test that ensures substrings of rope don't use too much memory.)

* JSTests/stress/many-substrings-of-rope-shouldnt-use-excessive-memory-2.js:
(test):
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::Heap):
(JSC::Heap::runEndPhase):
(JSC::Heap::projectedGCRateLimitingValue):
(JSC::Heap::updateAllocationLimits):
(JSC::Heap::collectIfNecessaryOrDefer):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Tools/Scripts/run-jsc-stress-tests:

Canonical link: https://commits.webkit.org/297631@main
